src/paulscode/android/mupen64plusae/profile/TouchscreenProfileActivity.java

```diff
@@ -385,37 +397,120 @@ private void hideSystemBars()
         int x = (int) event.getX();
         int y = (int) event.getY();
         
-        // Get the N64 index of the button that was pressed
-        int index = mTouchscreenMap.getButtonPress( x, y );
-        if( index != TouchMap.UNMAPPED )
+        if( ( event.getAction() & MotionEvent.ACTION_MASK ) == MotionEvent.ACTION_DOWN )
         {
-            String assetName = TouchMap.ASSET_NAMES.get( index );
-            String title = READABLE_NAMES.get( index );
+            initialX = x;
+            initialY = y;
+            dragIndex = TouchMap.UNMAPPED;
+            dragging = false;
             
-            // D-pad buttons are not holdable
-            if( DPAD.equals( assetName ) )
-                index = -1;
+            // ignore touches that start at the very bottom or top of the screen
+            View view = getWindow().getDecorView();
+            if (y < 10 || y > view.getHeight() - 10)
+                return false;
             
-            popupDialog( assetName, title, index );
+            // Get the N64 index of the button that was pressed
+            int index = mTouchscreenMap.getButtonPress( x, y );
+            if( index != TouchMap.UNMAPPED )
+            {
+                dragIndex = index;
+                dragAsset = TouchMap.ASSET_NAMES.get( index );
+                dragFrame = mTouchscreenMap.getButtonFrame( dragAsset );
+            }
+            else
+            {
+                // See if analog was pressed
+                Point point = mTouchscreenMap.getAnalogDisplacement( x, y );
+                int dX = point.x;
+                int dY = point.y;
+                float displacement = FloatMath.sqrt( ( dX * dX ) + ( dY * dY ) );
+                if( mTouchscreenMap.isInCaptureRange( displacement ) )
+                {
+                    dragIndex = -2;
+                    dragAsset = ANALOG;
+                    dragFrame = mTouchscreenMap.getAnalogFrame();
+                }
+            }
+            
+            dragX = mProfile.getInt( dragAsset + TAG_X, INITIAL_ASSET_POS );
+            dragY = mProfile.getInt( dragAsset + TAG_Y, INITIAL_ASSET_POS );
+            
+            return true;
         }
-        else
+        else if( ( event.getAction() & MotionEvent.ACTION_MASK ) == MotionEvent.ACTION_MOVE )
+        {
+            if (dragIndex != TouchMap.UNMAPPED)
+            {
+                if (!dragging)
+                {
+                    int dX = x - initialX;
+                    int dY = y - initialY;
+                    float displacement = FloatMath.sqrt( ( dX * dX ) + ( dY * dY ) );
+                    if (displacement >= 10)
+                        dragging = true;
+                }
+                if (!dragging)
+                    return false;
+                
+                // drag this button or analog stick around
+                
+                // calculate the X and Y percentage
+                View view = getWindow().getDecorView();
```

Comment:
Is this right? The values used in TouchMap.resize always seem to equal the size of the decorView, but maybe not? Not sure if I should be polling for these values from somewhere else.